crates/ruff_db/src/program.rs

```diff
@@ -64,6 +64,24 @@ impl std::fmt::Debug for TargetVersion {
     }
 }
 
+/// Configuration options for `site-packages` search paths.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum SitePackages {
+    /// Derive the `site-packages` paths from a path to a Python executable
+    Derived { venv_path: SystemPathBuf },
+
+    /// Pretend that all `site-packages` paths are known before the red-knot process starts.
+    ///
+    /// This isn't particularly realistic, but it's useful for mocking out site-packages in tests.
+    Known(Vec<SystemPathBuf>),
+}
+
+impl SitePackages {
+    pub fn none() -> Self {
+        Self::Known(vec![])
+    }
+}
```

Comment:
The paths in `Program` all must be resolved or it will be impossible to implement persistent caching. That means, any python and site package discovery **must** happen before setting the search paths on `Program`. 

That's why I think we don't need this enum.

Reply:
That makes sense to me, but then I don't understand why we're setting `SystemPathBuf`s (which represent unvalidated, unresolved search paths) on `Program` at all, rather than `ModulePath`s (which represent validated, resolved search paths). This was my concern I expressed in my review comment https://github.com/astral-sh/ruff/pull/12318#discussion_r1677156390 when you created the `Program` struct.

Reply:
I'm working on an alternative solution which is a more significant refactor, but perhaps will address your concern here

Reply:
We could validate the settings earlier, but we don't necessarily have to.  

The problem with storing `SearchPath` on `Program` is that `Program` is in `ruff_db`... That's what I commented in your other PR around validation.

Reply:
> I'm working on an alternative solution which is a more significant refactor, but perhaps will address your concern here

I think what we have here is fine, except that it can always be a vec. I don't think we have to do any significant refactors.

Reply:
The reason why I think we should do the validation outside of salsa is because I consider failing to resolve Python a fatal error. See this comment https://github.com/astral-sh/ruff/pull/12376#issuecomment-2247800148

That's why I want to avoid making any changes that need to be undone when migrating tho this model.

Reply:
> I consider failing to resolve Python a fatal error

Failing to resolve Python is a very different thing to failing to resolve the site-packages directories from a resolved path to a Python executable.

> I also strongly prefer avoiding any refactors until we've figured out the error handling. We should instead focus on shipping the features that we've set out for this week.

Fair enough, but this is one of the features that I set out for the week, that we agreed a couple of weeks ago that it would be good for me to work on 😅

I'm okay with putting off larger refactors for now, but it's unclear to me how to complete work on deriving site-packages from a Python executable unless we're agreed on where that logic should be. Anyway, I can do other work for now, and we can discuss in Runes planning tomorrow.

I did some interesting work today experimenting with broader refactors — I'm very unsure what I have locally is the way we want to go, but I think it was interesting anyway. I'll put it up tomorrow so you can see.

Reply:
> Failing to resolve Python is a very different thing to failing to resolve the site-packages directories from a resolved path to a Python executable.

What would you expect to happen if we fail to detect python (and its site packages) when running the CLI?

Reply:
That makes sense. For now, I would suggest to plug your logic somewhere into the red knot crate (or at least call it from there when constructing Program). I think the change from one to many site packages is still worth landing